File: backend-legacy-c/src/dice_mechanics.c
/*
 * dice_mechanics.c — Phase 0: Dice enumeration & probability
 *
 * Implements dice multiset operations on R_{5,6} (the 252 sorted 5-dice
 * multisets) and probability calculations P(⊥ → r).
 *
 * See: theory/optimal_yahtzee_pseudocode.md, Phase 0: PRECOMPUTE_ROLLS_AND_PROBABILITIES.
 */
#include <stdlib.h>
#include <math.h>

#include <context.h>

/* Simulation helper: roll all 5 dice uniformly at random. */
void RollDice(int dice[5]) {
    for (int i = 0; i < 5; i++) {
        dice[i] = rand() % 6 + 1;
    }
}

/* Simulation helper: reroll dice indicated by mask (bit i = reroll die i). */
void RerollDice(int dice[5], const int mask) {
    for (int i = 0; i < 5; i++) {
        if (mask & (1 << i)) {
            dice[i] = rand() % 6 + 1;
        }
    }
}

/* Normalize dice to canonical sorted form (ascending).
 * All dice sets in R_{5,6} are stored sorted for deduplication. */
void SortDiceSet(int arr[5]) {
    for (int i = 0; i < 4; i++) {
        for (int j = i + 1; j < 5; j++) {
            if (arr[j] < arr[i]) {
                const int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
    }
}

/* Map a sorted dice set to its index in R_{5,6} (0–251).
 * Uses the precomputed 5D index_lookup table for O(1) access. */
int FindDiceSetIndex(const YatzyContext *ctx, const int dice[5]) {
    return ctx->index_lookup[dice[0] - 1][dice[1] - 1][dice[2] - 1][dice[3] - 1][dice[4] - 1];
}

/* Compute P(⊥ → r): probability of rolling this exact sorted dice set
 * from 5 fresh dice.
 *
 * Formula: multinomial(5; face_counts) / 6^5
 *        = 5! / (n1! · n2! · ... · n6!) / 6^5
 *
 * See pseudocode Phase 0: "For each full roll r in R_5,6:
 *   Compute P(⊥ → r) = multinomial(5; counts of each face in r) / 6^5" */
double ComputeProbabilityOfDiceSet(const YatzyContext *ctx, const int dice[5]) {
    int face_count[7];
    for (int i = 1; i <= 6; i++) face_count[i] = 0;
    for (int i = 0; i < 5; i++) face_count[dice[i]]++;

    const int numerator = ctx->factorial[5];
    int denominator = 1;
    for (int f = 1; f <= 6; f++) {
        if (face_count[f] > 1) {
            const int fc = face_count[f];
            int sub_fact = 1;
            for (int x = 2; x <= fc; x++) sub_fact *= x;
            denominator *= sub_fact;
        }
    }

    const double permutations = (double) numerator / (double) denominator;
    const double total_outcomes = pow(6, 5);
    return permutations / total_outcomes;
}

/* Count occurrences of each face (1–6) in a 5-dice set.
 * face_count[0] is unused; face_count[f] = count of face f. */
void CountFaces(const int dice[5], int face_count[7]) {
    for (int i = 1; i <= 6; i++) face_count[i] = 0;
    for (int i = 0; i < 5; i++) face_count[dice[i]]++;
}

/* Scoring helper for N-of-a-kind categories.
 * Returns highest_face * n if any face appears >= n times, else 0. */
int NOfAKindScore(const int face_count[7], int n) {
    for (int face = 6; face >= 1; face--) {
        if (face_count[face] >= n) return face * n;
    }
    return 0;
}

File: backend-rust/src/dice_mechanics.rs
use crate::types::YatzyContext;

/// Count occurrences of each face (1-6) in a 5-dice set.
/// face_count[0] is unused; face_count[f] = count of face f.
pub fn count_faces(dice: &[i32; 5]) -> [i32; 7] {
    let mut face_count = [0i32; 7];
    for &d in dice {
        face_count[d as usize] += 1;
    }
    face_count
}

/// Normalize dice to canonical sorted form (ascending).
pub fn sort_dice_set(arr: &mut [i32; 5]) {
    for i in 0..4 {
        for j in (i + 1)..5 {
            if arr[j] < arr[i] {
                arr.swap(i, j);
            }
        }
    }
}

/// Map a sorted dice set to its index in R_{5,6} (0-251).
#[inline(always)]
pub fn find_dice_set_index(ctx: &YatzyContext, dice: &[i32; 5]) -> usize {
    ctx.index_lookup[(dice[0] - 1) as usize][(dice[1] - 1) as usize][(dice[2] - 1) as usize]
        [(dice[3] - 1) as usize][(dice[4] - 1) as usize] as usize
}

/// Compute P(empty -> r): probability of rolling this exact sorted dice set
/// from 5 fresh dice.
///
/// Formula: multinomial(5; face_counts) / 6^5
///        = 5! / (n1! * n2! * ... * n6!) / 6^5
pub fn compute_probability_of_dice_set(ctx: &YatzyContext, dice: &[i32; 5]) -> f64 {
    let mut face_count = [0i32; 7];
    for &d in dice {
        face_count[d as usize] += 1;
    }

    let numerator = ctx.factorial[5] as f64;
    let mut denominator = 1.0;
    for f in 1..=6 {
        if face_count[f] > 1 {
            let mut sub_fact = 1;
            for x in 2..=face_count[f] {
                sub_fact *= x;
            }
            denominator *= sub_fact as f64;
        }
    }

    let permutations = numerator / denominator;
    let total_outcomes = 6.0f64.powi(5);
    permutations / total_outcomes
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::phase0_tables;

    fn make_ctx() -> Box<YatzyContext> {
        let mut ctx = Box::new(YatzyContext::new());
        phase0_tables::precompute_lookup_tables(&mut ctx);
        ctx
    }

    #[test]
    fn test_sort_dice_set() {
        let mut d1 = [5, 3, 1, 4, 2];
        sort_dice_set(&mut d1);
        assert_eq!(d1, [1, 2, 3, 4, 5]);

        let mut d2 = [1, 2, 3, 4, 5];
        sort_dice_set(&mut d2);
        assert_eq!(d2, [1, 2, 3, 4, 5]);

        let mut d3 = [6, 5, 4, 3, 2];
        sort_dice_set(&mut d3);
        assert_eq!(d3, [2, 3, 4, 5, 6]);

        let mut d4 = [3, 3, 3, 3, 3];
        sort_dice_set(&mut d4);
        assert_eq!(d4, [3, 3, 3, 3, 3]);
    }

    #[test]
    fn test_find_dice_set_index() {
        let ctx = make_ctx();
        assert_eq!(find_dice_set_index(&ctx, &[1, 1, 1, 1, 1]), 0);
        assert_eq!(find_dice_set_index(&ctx, &[6, 6, 6, 6, 6]), 251);

        // Round-trip
        for i in 0..252 {
            assert_eq!(find_dice_set_index(&ctx, &ctx.all_dice_sets[i]), i);
        }
    }

    #[test]
    fn test_count_faces() {
        let fc = count_faces(&[1, 1, 2, 3, 3]);
        assert_eq!(fc[1], 2);
        assert_eq!(fc[2], 1);
        assert_eq!(fc[3], 2);
        assert_eq!(fc[4], 0);
        assert_eq!(fc[5], 0);
        assert_eq!(fc[6], 0);

        let fc2 = count_faces(&[6, 6, 6, 6, 6]);
        assert_eq!(fc2[6], 5);
        assert_eq!(fc2[1], 0);
    }

    #[test]
    fn test_probability() {
        let ctx = make_ctx();

        let p1 = compute_probability_of_dice_set(&ctx, &[1, 1, 1, 1, 1]);
        assert!((p1 - 1.0 / 7776.0).abs() < 1e-12);

        let p2 = compute_probability_of_dice_set(&ctx, &[1, 1, 1, 1, 2]);
        assert!((p2 - 5.0 / 7776.0).abs() < 1e-12);

        let sum: f64 = ctx.dice_set_probabilities.iter().sum();
        assert!((sum - 1.0).abs() < 1e-9);
    }
}
